rfd/0022-ssh-agent-forwarding.md

```diff
@@ -0,0 +1,224 @@
+        │                      │               ║ User connects to ║               │
+        │                      │               ║ "Third Machine"  ║               │
+        │                      │               ╚════════╤═════════╝               │
+        │                      │                        │       SSH Connect       │
+        │                      │                        │ ───────────────────────>│
+        │                      │                        │     Auth challenge      │
+        │                      │                        │ <───────────────────────│
+        │  Auth Challenge (via forwarded connection)    │                         │
+        │<──────────────────────────────────────────────│                         │
+        │  Signed Response (via forwarded connection)   │                         │
+        │──────────────────────────────────────────────>│                         │
+        │                      │                        │     Signed Response     │
+        │                      │                        │ ───────────────────────>│
+        │                      │                ╔═══════╧═════════════════════════╧═══════╗
+        │                      │                ║       Authentication established        ║
+        |                      |                ╚═════════════════════════════════════════╝
+   ┌────┴────┐          ┌──────┴──────┐          ┌──────┴───────┐          ┌──────┴──────┐
+   │Key Agent│          │Local Machine│          │Remote Machine│          │Third Machine│
+   └─────────┘          └─────────────┘          └──────────────┘          └─────────────┘
+```
+
+The above diagram leaves out a _lot_ of detail, but the overall process should be clear
+enough for this discussion.
+
+### What's the issue?
+
+An OpenSSH user can forward their Key Agent to the remote machine using the `-A` flag on
+the `ssh` command line. This allows the user to use arbitrary keys stored on their _local_
+machine to authenticate against a "Third Machine" from inside their session on the Remote
+Machine.
+
+When running `tsh ssh`, a user may have _two_ independent key agents running: their own
+(which we will call the _User Key Agent_), and an ephemeral one inside the `tsh` process
+itself (the _Teleport Key Agent_). The Teleport Key Agent is populated with the contents
+of the user's `.tsh` directory.
+
+The `tsh ssh` client _also_ offers a `-A` option to forward a Key Agent to the Remote
+Machine, but it will only ever forward the Teleport Key Agent, not the User Key Agent.
+This is surprising behaviour to a user accustomed to OpenSSH. The user finds that keys
+they expected to be available to them in the session on the Remote Machine are not
+available.
+
+The tricky part is that _both_ of these behaviours are sensible, depending on the user's
+expectations. Any solution will have to accommodate both behaviours.
+
+## Details
+
+### Proposed Solution
```

Comment:
Can you add a section to make it clear that this does not affect auth (which will still use keys on disk), these changes are specific to agent forwarding.

Reply:
Done